### DISCUSSION TITLE
Visual Studio Code debugs python 2.7 app in a docker container with python3

### DISCUSSION THREAD

**Discussion by LeiteMarga** [microsoft/vscode-docker#3585] [Q&A] [answered]
Hi there, 

I am having some trouble getting my debugger to work in VS Code and from the error messages below, I gather that there is a discrepancy between the python versions.  The legacy app I am working on is using Python 2.7.

I have come across this similar issue: [Issue reference](https://stackoverflow.com/questions/69360649/visual-studio-code-debugs-python-2-7-app-in-a-docker-container-with-python3) , however still can't get it to work.  

Is there anything obvious that I'm missing?  Any assistance will be greatly appreciated.

- I'm using a Mac OS

**Pop up Error 1:**
Invalid message: "debugLauncherPath" must be str

**Terminal error:** 
docker run -dt -P --name "app_name-dev" --label "com.microsoft.created-by=visual-studio-code" -v "/User/.vscode/extensions/ms-python.python-2021.10.1365161279/pythonFiles/lib/python/debugpy:/debugpy:ro,z" --entrypoint "python3" "app_name:latest" <

### My configurations are: 
**launch.json**
{
    "version"       : "0.2.0",
    "configurations": [
        {
            "name": "Docker: Python - Django",
            "type": "docker",
            "request": "launch",
            "preLaunchTask": "docker-run: debug",
            "python": {
                "pathMappings": [
                    {
                        "localRoot": "${workspaceFolder}",
                        "remoteRoot": "/app"
                    }
                ],
                "projectType": "django",
            },
        },
    ]
}

**tasks.json:**
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "docker-build",
			"label": "docker-build",
			"platform": "python",
			"dockerBuild": {
				"tag": "app_name:latest",
				"dockerfile": "${workspaceFolder}/docker/DockerFileLocal",
				"context": "${workspaceFolder}",
				"pull": true
			}
		},
		{
			"type": "docker-run",
			"label": "docker-run: debug",
			"dependsOn": [
				"docker-build"
			],
			"python": {
				"args": [
					"runserver",
					"0.0.0.0:8000",
					"--nothreading",
					"--noreload"
				],
				"file": "manage.py"
			}
		}
	]
}

**settings.py**
{
    "workbench.colorTheme": "One Dark Pro",
    "git.autofetch": true,
    "[jsonc]": {
        "editor.defaultFormatter": "vscode.json-language-features"
    },
    "[html]": {
        "editor.defaultFormatter": "esbenp.prettier-vscode"
    },
    
    "files.exclude": {
        "**/__pycache__": true,
        "**/*.pyc"      : true
        },

    "[python]": {

        "editor.rulers": [79]},
        "editor.multiCursorModifier"     : "alt",
        "python.languageServer"          : "Jedi",
        "python.analysis.cacheFolderPath": "lint",
        "python.defaultInterpreterPath"  : "~/.pyenv/versions/2.7.13/envs/app_name/bin/python",
        "python.linting.enabled"         : true,
        "python.linting.pylintEnabled"   : true,
        "python.linting.flake8Enabled"   : true,
        "python.linting.flake8Path"      : "~/.pyenv/versions/2.7.13/envs/app_name/bin/flake8",
        "python.linting.flake8Args"      : ["--config=pep8_test","APP/"],
        "typescript.validate.enable"     : false,
        "javascript.validate.enable"     : false,
        "gitlens.blame.format"           : "${author}${agoOrDate|14-}",
        "git.confirmSync"                : false,
        "redhat.telemetry.enabled"       : false,
        "files.autoSave"                 : "afterDelay",
}

**manage.py**
import os
import sys

if __name__ == "__main__":
    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "APP.settings")

    from django.core.management import execute_from_command_line

    execute_from_command_line(sys.argv)

**Comment by bwateratmsft** [comment]
It looks like after one of your edits the terminal error message isn't there, can you share it?

**Reply by LeiteMarga** [reply]
Yes, I edited out one of the errors that I fixed after my original post.  This is the updated terminal error log: 

> Executing task: docker-run: debug <

> docker run -dt -P --name "app-name-dev" --label "com.microsoft.created-by=visual-studio-code" -v "/Users/margaleite/.vscode/extensions/ms-python.python-2021.10.1365161279/pythonFiles/lib/python/debugpy:/debugpy:ro,z" --entrypoint "python3" "app-name:latest" <

f278998e488d1be3880ec1f095c60d362c3501cac9be399dc9635f92abb098fc

Terminal will be reused by tasks, press any key to close it.

**Reply by LeiteMarga** [reply]
In the related issue that I referenced, they mention adding the below to the Python entry point file.  I assume this refers to the manage.py file in Django, but couldn't get it to work.  Is the entry point file perhaps somewhere else?

*I have ptvsd (4.3.2) installed, but in the current state am not importing it anywhere as per the below code from the related issue:

import ptvsd
ptvsd.enable_attach(address=('0.0.0.0', 5678))

**Reply by bwateratmsft** [reply]
Aha, I missed this part when I read your post the first time:
> The legacy app I am working on is using Python 2.7.

Obviously I have to recommend working on migrating the app to Python 3+, as Python 2 has been end-of-life for a couple of years now.

`ptvsd` is no longer used for debugging Python apps, now `debugpy` is used. I'm not sure if `debugpy` is compatible with Python 2.7. @int19h is there a way to still use `ptvsd` for debugging? Or is the best bet here to install an older version of the Docker and Python extensions?

**Reply by LeiteMarga** [reply]
Hi Brandon,  yes it's definitely high priority for us to upgrade the system in the next couple of months, however I am new to the project and it will help a lot if I can get the debugger to work in the meantime as well.

**Reply by bwateratmsft** [reply]
@LeiteMarga One thing that might work is to use a Python attach configuration instead of direct launch.

1. Add a Python: Remote Attach launch configuration:
```json
        {
            "name": "Python: Remote Attach",
            "type": "python",
            "request": "attach",
            "connect": {
                "host": "localhost",
                "port": 5678
            },
            "pathMappings": [
                {
                    "localRoot": "${workspaceFolder}",
                    "remoteRoot": "."
                }
            ],
            "justMyCode": true
        }
```
2. Right click the `docker-compose-debug.yml` file added by Dockerfile scaffolding, and do "Compose Up"
3. Select the remote attach configuration and hit <kbd>F5</kbd>

**Comment by int19h** [comment]
debugpy is no longer compatible with anything earlier than Python 3.7. However, earlier versions of debugpy were compatible with Python 2.7, and can still be used for "attach" / `debugpy.listen()` debugging. If you do `pip install debugpy` from a Python 2.7 environment, that should get you the most recent version that is still compatible.

**Comment by LeiteMarga** [comment, marked as answer]
Thank you so much for the feedback - really appreciate it!  We managed to get it to work using ptvsd with the following configurations:

**manage.py**
from django.conf import settings

    if settings.DEBUG:
        print(os.environ.get('RUN_MAIN') or os.environ.get('WERKZEUG_RUN_MAIN'))
        if os.environ.get('RUN_MAIN') or os.environ.get('WERKZEUG_RUN_MAIN'):
            import ptvsd
            ptvsd.enable_attach(address = ('0.0.0.0', 3000))
            print "Attached remote debugger"

**launch.json**
{
    "name": "Django",
    "type": "python",
    "request": "attach",
    "pathMappings": [
        {
            "localRoot": "${workspaceFolder}",
            "remoteRoot": "/app"
        }
    ],
    "port": 3000,
    "host": "localhost"
}

**docker-compose.yml**
ports:
      - "8000:8000"
      - "3000:3000"

**Reply by bwateratmsft** [reply]
Great news! Glad it's working now.
